### DISCUSSION TITLE
How to link Android aar into C++ code

### DISCUSSION THREAD

**Discussion by ospfranco** [trustwallet/wallet-core#2074] [Q&A]
Hey, I'm trying to write an Android C++ library using wallet-core, however, I'm a C++/Android noob. I've managed to add the Gradle dependency into the Android project. Then I would like to expose the internal C++ headers to my own C++ code.

To write my own library I'm using CMake (with a CMakeLists.txt file) where I just declare where my files are. I got from the documentation it is possible for packaged libraries (.aar) to expose the internal C++ headers to other C++ code. I basically have the Gradle dependency working and now the question is how to expose/link it to my C++ code.

```Gradle
// My library's build.gradle
android {
// From what I get prefab needs to be turned on for .aar libraries to expose the C++ headers
  buildFeatures {
    prefab true
  }

// The maven repo is configured somewhere else and the dependency is correctly downloaded
  dependencies {
    implementation 'com.trustwallet:wallet-core:2.7.2'
  }

}
```
How ever when I try to add the dependency on my `CMakeLists.txt` then I just get an error, here is my complete CMakeLists file:

```CMake
cmake_minimum_required(VERSION 3.9.0)

set (PACKAGE_NAME "react-native-wallet-core")
set (BUILD_DIR ./build)

include_directories(
        ${PACKAGE_NAME}
        "${NODE_MODULES_DIR}/react-native/React"
        "${NODE_MODULES_DIR}/react-native/React/Base"
        "${NODE_MODULES_DIR}/react-native/ReactCommon/jsi"
        "../cpp"
        "."
)

file (GLOB LIBRN_DIR "${BUILD_DIR}/react-native-0*/jni/${ANDROID_ABI}")
message(STATUS "LIBRN_DIR: ${LIBRN_DIR}")

if (EXISTS "${LIBRN_DIR}/libjsi.so")
  # React Native 0.66.x and above

  find_library(
          JSI_LIB
          jsi
          PATHS ${LIBRN_DIR}
          NO_CMAKE_FIND_ROOT_PATH
  )

  add_library(
          ${PACKAGE_NAME}
          SHARED
          ../cpp/react-native-wallet-core.cpp
          ../cpp/react-native-wallet-core.h
          ./cpp-adapter.cpp
  )
else()
  # React Native 0.65.x and below
  set (JSI_LIB "")
  add_library(
          ${PACKAGE_NAME}
          SHARED
          "${NODE_MODULES_DIR}/react-native/ReactCommon/jsi/jsi/jsi.cpp"
          ../cpp/react-native-wallet-core.cpp
          ../cpp/react-native-wallet-core.h
          ./cpp-adapter.cpp
  )
endif()

find_library(
        LOG_LIB
        log
)

find_library(
        REACT_NATIVE_JNI_LIB
        reactnativejni
        PATHS ${LIBRN_DIR}
        NO_CMAKE_FIND_ROOT_PATH
)

find_library(
        TRUST_WALLET_CORE
        TrustWalletCore
)

set_target_properties(
        ${PACKAGE_NAME} PROPERTIES
        CXX_STANDARD 17
        CXX_EXTENSIONS OFF
        POSITION_INDEPENDENT_CODE ON
)

target_link_libraries(
        ${PACKAGE_NAME}
        ${LOG_LIB}
        ${REACT_NATIVE_JNI_LIB}
        ${JSI_LIB}
        ${TRUST_WALLET_CORE}
        android
)
```

You can see I try to find the `TrustWalletCore` library and link it to my library. But the gradle process just fails:

```
FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':reactnativewalletcore:generateJsonModelDebug'.
> /Users/osp/Developer/jsi-rn-wallet-core/android/CMakeLists.txt : C/C++ debug|armeabi-v7a : CMake Error: The following variables are used in this project, but they are set to NOTFOUND.
  Please set them or make sure they are set and tested correctly in the CMake files:
  TRUST_WALLET_CORE
      linked by target "react-native-wallet-core" in directory /Users/osp/Developer/jsi-rn-wallet-core/android
```

Are there any pointers at all you could give me? is it my fault? or is it because the generated `.aar` is not `prefab` compatible and importing the code into C++ won't work?

**Comment by hewigovens** [comment]
Currently android build is built as dynamic library, I don't think we configured `prefab`. From the cmake error, where is `TRUST_WALLET_CORE` defined? one easy option is to only implement a Java/Kotlin wrapper on top of wallet core

**Comment by ospfranco** [comment]
Thanks for replying, indeed, it seems I cannot just turn on prefab, the generated `.aar` needs to be packaged with the prefab plugin enabled. I and my team are now trying to compile the project on our own with `prefab`, any tips you could give to make this work?

Regarding the CMake error, I think it is not important until we get the library compiled. Also, using Java/Kotlin is not an option, since I'm trying to create a RN library that using JSI (faster C++ bindings with JS) and I want to use a single codebase, hence I need the C++ headers.

**Reply by hewigovens** [reply]
I don't know `prefab` maybe @vcoolish could help?

**Comment by ospfranco** [comment]
I've managed to compile a prefab .aar (all the headers seem to be there), however when I try to compile my app, I get this error:

```
Error: Command failed: ./gradlew app:installDebug -PreactNativeDevServerPort=8081

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':reactnativewalletcore:generateJsonModelDebug'.
> error when building with cmake using /Users/osp/Developer/jsi-rn-wallet-core/android/CMakeLists.txt: Build command failed.
  Error while executing java process with main class com.google.prefab.cli.AppKt with arguments {--build-system cmake --platform android --abi armeabi-v7a --os-version 23 --stl c++_static --ndk-version 21 --output /Users/osp/Developer/jsi-rn-wallet-core/android/.cxx/cmake/debug/prefab/armeabi-v7a/prefab /Users/osp/.gradle/caches/transforms-3/b5eeafd8e58afb07fc74ac92d29f5be6/transformed/jetified-trustwalletcore-release/prefab}
  
  No compatible library found for //trustwalletcore/TrustWalletCore. Rejected the following libraries:
  android.arm64-v8a: User is targeting armeabi-v7a but library is for arm64-v8a
  android.armeabi-v7a: Library is a shared library with a statically linked STL and cannot be used with any library using the STL
  android.x86: User is targeting armeabi-v7a but library is for x86
  android.x86_64: User is targeting armeabi-v7a but library is for x86_64
```

The `Library is a shared library with a statically linked STL and cannot be used with any library using the STL` error seems to be the most important one, I googled a bit and tried changing stl link to static but that doesn't seem to work either:

https://github.com/google/oboe/issues/812

Any guidance would be much appreciated

**Reply by vcoolish** [reply]
you need to compile it with targeting architectures, add this to your gradle file
include 'x86', 'x86_64', 'armeabi-v7a', 'arm64-v8a'

**Reply by vcoolish** [reply]
it's written in error, you are targeting armeabi-v7a only and your lib is compiled for arm

**Reply by ospfranco** [reply]
My library's `build.gradle` already includes the ABI filters, should I add the `include` again?

<img width="655" alt="CleanShot 2022-03-16 at 12 44 19@2x" src="https://user-images.githubusercontent.com/1634213/158582796-4b2d2043-eaa8-49c2-b20d-49a195624bf3.png">

**Reply by AlexOnomy** [reply]
There are multiple errors listed because it is explaining what libraries are not being used and there are multiple builds.

This is the right build for the right plataform and has a different error from the rest.

android.armeabi-v7a: Library is a shared library with a statically linked STL and cannot be used with any library using the STL

**Comment by ospfranco** [comment]
I've been reading up on the Android STL and static/dynamic linking, so it would seem when the `aar` is generated, it is generated using a statically linked STL.

The documentation says changing `ANDROID_STL` flag, should either link against the shared STL or an NDK specific static one. I tried that:

<img width="685" alt="CleanShot_2022-03-17_at_09 21 412x" src="https://user-images.githubusercontent.com/1634213/158772145-2a3efbcb-4aa3-4073-842a-8447229ad8a3.png">

But I still get the same error when trying to link it in my library. As @AlexOnomy mentioned, it is not a problem with the missing arch missing, but rather a specific version linking error.
